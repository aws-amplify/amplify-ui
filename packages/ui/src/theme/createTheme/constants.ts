export const DEFAULTS = {
  ignoreKeys: ['original'],
};

export const OPTIONS = {
  openingCharacter: '{',
  closingCharacter: '}',
  separator: '.',
};

export const REFERENCE_REGEX = /\{([^}]+)\}/g;
