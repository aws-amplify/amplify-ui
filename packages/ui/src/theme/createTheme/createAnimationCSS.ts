import { isFunction } from '../../utils';
import { CSSProperties } from '../components/utils';
import { WebTokens } from '../tokens';
import { Animations } from '../types';
import { propsToString } from './utils';

export function createAnimationCSS({
  animations,
  tokens,
}: {
  animations: Animations;
  tokens: WebTokens;
}) {
  let cssText = '';
  Object.entries(animations).forEach(([key, value]) => {
    cssText += `\n  @keyframes ${key} {`;
    Object.entries(value).forEach(([step, properties]) => {
      cssText += `\n    ${step} {\n`;
      const animationProperties: CSSProperties = isFunction(properties)
        ? (properties(tokens) as CSSProperties)
        : properties;
      cssText += propsToString(animationProperties);
      cssText += `\n    }`;
    });
    cssText += `\n  }`;
  });
  return cssText;
}
