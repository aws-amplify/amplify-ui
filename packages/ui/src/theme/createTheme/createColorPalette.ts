import { ColorValues, ScaleKey } from '../tokens/colors';

/**
 * Takes a set of keys and a color name and returns an object of design tokens,
 * used for applying a primary color at the theme level to our tokens.
 *
 * createColorPalette({keys: ['10','20',...], value: 'red'})
 * returns {
 *   10: { value: '{colors.red.10.value}' },
 *   20: { value: '{colors.red.20.value}' },
 *   ...
 * }
 */
export function createColorPalette<
  ColorType extends ColorValues<ScaleKey, 'default'> = ColorValues<
    ScaleKey,
    'default'
  >,
>({ keys, value }: { keys: string[]; value: string }): ColorType {
  return keys.reduce((acc, key) => {
    return {
      ...acc,
      [key]: { value: `{colors.${value}.${key}.value}` },
    };
  }, {} as ColorType);
}
