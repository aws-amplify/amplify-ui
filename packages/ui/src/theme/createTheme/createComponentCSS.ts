import { DefaultTheme, WebTheme } from '../types';
import { propsToString } from './utils';
import { ComponentsTheme } from '../components';
import { isFunction, splitObject } from '../../utils';
import { BaseTheme } from '../components/utils';

function addVars(selector: string, vars: Record<string, unknown>) {
  if (!vars) return '';

  return `${selector} { ${Object.entries(vars)
    .map(([key, value]) => {
      return `--${key}:${value}; `;
    })
    .join(' ')}}\n`;
}

function recursiveComponentCSS(baseSelector: string, theme: BaseTheme) {
  let str = '';
  const { _modifiers = {}, _element = {}, _vars, ...props } = theme;

  // if there are no props, skip
  if (Object.keys(props).length) {
    // separate psuedo/attribute selectors
    const [selectors, other] = splitObject(
      props,
      (key) => key.startsWith(':') || key.startsWith('[')
    );

    Object.entries(selectors).forEach(([selector, value]) => {
      // need to remove nested things like vars and elements
      const {
        _modifiers = {},
        _element = {},
        _vars,
        ...props
      } = value as BaseTheme;
      str += `${baseSelector}${selector} { ${propsToString(props)} }\n`;
      str += addVars(`${baseSelector}${selector}`, _vars);
    });

    str += `${baseSelector} { ${propsToString(other)} }\n`;
  }

  str += addVars(baseSelector, _vars);

  Object.entries(_modifiers).forEach(([key, value]) => {
    if (value && Object.keys(value).length) {
      str += recursiveComponentCSS(`${baseSelector}--${key}`, value);
    }
  });

  Object.entries(_element).forEach(([key, value]) => {
    if (value && Object.keys(value).length) {
      str += recursiveComponentCSS(`${baseSelector}__${key}`, value);
    }
  });

  return str;
}

/**
 * This will take a component theme and create the appropriate CSS for it.
 *
 */
export function createComponentCSS(
  themeName: string,
  components: Array<ComponentsTheme>,
  tokens: WebTheme['tokens'],
  breakpoints: DefaultTheme['breakpoints']
) {
  let cssText = '';

  components.forEach(({ name, theme, overrides }) => {
    const baseComponentClassName = `amplify-${name}`;
    const componentClassName = `[data-amplify-theme="${themeName}"] .${baseComponentClassName}`;
    // unwrap the component theme
    // if it is a function: call it with the defaultTheme to get a static object
    const componentTheme: BaseTheme = isFunction(theme)
      ? (theme(tokens) as BaseTheme)
      : theme;

    cssText += recursiveComponentCSS(componentClassName, componentTheme);

    // if the component theme has overrides
    // generate the appropriate CSS for each of them
    if (overrides) {
      overrides.forEach((override) => {
        // unwrap the override component theme just like above
        const componentTheme: BaseTheme = isFunction(override.theme)
          ? (override.theme(tokens) as BaseTheme)
          : override.theme;

        if ('mediaQuery' in override) {
          cssText += `@media (${
            override.mediaQuery
          }) {\n ${recursiveComponentCSS(
            componentClassName,
            componentTheme
          )} \n}`;
        }
        if ('breakpoint' in override) {
          const breakpoint = breakpoints.values[override.breakpoint];
          cssText += `\n@media (min-width: ${breakpoint}px) {\n ${recursiveComponentCSS(
            componentClassName,
            componentTheme
          )} \n}`;
        }
        if ('selector' in override) {
          cssText += recursiveComponentCSS(
            `${override.selector} .${baseComponentClassName}`,
            componentTheme
          );
        }
        if ('colorMode' in override) {
          cssText += `\n@media (prefers-color-scheme: ${
            override.colorMode
          }) {\n${recursiveComponentCSS(
            `[data-amplify-theme="${themeName}"][data-amplify-color-mode="system"] .${baseComponentClassName}`,
            componentTheme
          )}\n}\n`;
          cssText += recursiveComponentCSS(
            `[data-amplify-theme="${themeName}"][data-amplify-color-mode="${override.colorMode}"] .${baseComponentClassName}`,
            componentTheme
          );
        }
      });
    }
  });

  return cssText;
}
