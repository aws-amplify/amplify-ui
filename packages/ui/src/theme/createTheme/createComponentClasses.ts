import { isObject, isString, classNames, ClassNamesArgs } from '../../utils';
import { ComponentThemeFromName } from '../components';
import { BaseTheme } from '../components/utils';

// Gets the element names in a theme
type ElementNames<T extends unknown> = T extends { _element?: any }
  ? keyof Required<T['_element']>
  : never;

// Gets the modifiers of an element within a theme
type ModifierNames<T extends unknown> = T extends { _modifiers?: any }
  ? Arrayify<
      | keyof Required<T['_modifiers']>
      | {
          [key in keyof Required<T['_modifiers']>]?: boolean | null | undefined;
        }
      | undefined
    >
  : never;

type Arrayify<T> = T | T[];

// This type will take a component theme
// and turn it into the shape of the classname function
// for that component based on what is in the _element
// and _modifiers
export type ClassNameArgs<T extends BaseTheme> = {
  _element?:
    | ElementNames<Required<T>>
    | {
        [Key in ElementNames<Required<T>>]?: ModifierNames<
          Required<Required<Required<T>['_element']>[Key]>
        >;
      };
  _modifiers?: ModifierNames<Required<T>>;
};

export type ClassNameFunction<
  T extends BaseTheme = BaseTheme,
  NameType extends string = string,
> = (
  props?: ClassNameArgs<UnwrapTheme<ComponentThemeFromName<NameType, T>>>,
  extraClassnames?: ClassNamesArgs
) => string;

// This will take a theme which could be either a plain object or a function that returns a plain object
// and makes it the plain object
type UnwrapTheme<ThemeType extends BaseTheme = BaseTheme> = ThemeType extends (
  ...args: any
) => any
  ? ReturnType<ThemeType>
  : ThemeType;

export function createComponentClasses<
  ThemeType extends BaseTheme,
  NameType extends string = string,
>({ name, prefix = 'amplify-' }: { name: NameType; prefix?: string }) {
  const className: ClassNameFunction<ThemeType, NameType> = (
    props = {},
    extraClassnames = []
  ) => {
    const baseComponentClassName = `${prefix}${name}`;

    // get the element if there is one
    // the _element argument can be a string
    // like { _element: 'icon' }
    // or it could be an object where the key is
    // the element name and the value is the modifiers
    // like { _element: { icon: [size] } }
    const element = isString(props._element)
      ? props._element
      : isObject(props._element)
      ? Object.keys(props._element)[0]
      : undefined;

    const className = element
      ? `${baseComponentClassName}__${element}`
      : baseComponentClassName;
    const names = [className];

    if (element) {
      const modifiers = props._element[element];
      names.push(...modifierClassnames({ className, modifiers }));
    } else {
      names.push(
        ...modifierClassnames({
          className,
          modifiers: props._modifiers as ModifierNames<BaseTheme>,
        })
      );
    }

    return classNames([...names, ...extraClassnames]);
  };
  return className;
}

function modifierClassnames({
  className,
  modifiers,
}: {
  className: string;
  modifiers: ModifierNames<BaseTheme>;
}) {
  if (Array.isArray(modifiers)) {
    return modifiers.map((modifier) => {
      if (!modifier || !isString(modifier)) {
        return;
      }
      return `${className}--${modifier}`;
    });
  }
  if (isObject(modifiers)) {
    return Object.entries(modifiers).map(([key, value]) => {
      if (value) {
        return `${className}--${key}`;
      }
    });
  }
  if (isString(modifiers)) {
    return [`${className}--${modifiers}`];
  }
  return [];
}
