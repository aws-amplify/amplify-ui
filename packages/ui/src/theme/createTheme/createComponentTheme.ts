import { ComponentsTheme } from '../components';
import {
  BaseTheme,
  ComponentTheme,
  ComponentThemeOverride,
} from '../components/utils';
import { WebTokens } from '../tokens';
import {
  createComponentClasses,
  ClassNameFunction,
} from './createComponentClasses';

interface CreateComponentThemeProps<
  ThemeType extends BaseTheme,
  TokensType extends WebTokens,
  OverridesType extends BaseTheme = BaseTheme,
> {
  name: string;
  theme?: ComponentTheme<ThemeType, TokensType>;
  overrides?: ComponentThemeOverride<
    ComponentTheme<OverridesType, TokensType>
  >[];
  // overrides do not need to specify all the modifiers and elements
  // need to fix this
}

/**
 * Use this to create the theme of a component. You can use this
 * to both completely customize built-in components and
 * build your own components!
 *
 * @experimental
 *
 * ```ts
 * // built-in component styling
 * const alertTheme = createComponentTheme<AlertTheme>({
 *   name: 'alert',
 *   theme: (tokens) => {
 *     return {
 *       padding: tokens.space.large
 *     }
 *   }
 * });
 *
 * const theme = createTheme({
 *   name: 'my-theme',
 *   components: [alertTheme]
 * })
 * ```
 *
 * @param {Object} params
 * @param {string} params.name  - The name of the component. Use a built-in component name like button to theme buttons.
 * @returns
 */
export function createComponentTheme<
  TokensType extends WebTokens,
  ThemeType extends BaseTheme = BaseTheme,
>({
  name,
  theme,
  overrides,
}:
  | CreateComponentThemeProps<ThemeType, TokensType>
  | ComponentsTheme<TokensType>): {
  className: ClassNameFunction<ThemeType>;
  theme: typeof theme;
  name: string;
  overrides?: typeof overrides;
} {
  const prefix = 'amplify-';
  const className = createComponentClasses<ThemeType>({
    name,
    prefix,
  });
  return {
    className,
    theme,
    overrides,
    name,
  };
}
