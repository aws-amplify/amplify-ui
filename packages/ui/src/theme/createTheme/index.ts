export { createTheme } from './createTheme';
export { defineComponentTheme } from './defineComponentTheme';
export {
  cssNameTransform,
  setupTokens,
  SetupToken,
  isDesignToken,
} from './utils';
export {
  createComponentClasses,
  ClassNameArgs,
} from './createComponentClasses';
