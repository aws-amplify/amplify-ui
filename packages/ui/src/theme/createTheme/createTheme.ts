import { defaultTheme } from '../defaultTheme';
import { Theme, DefaultTheme, WebTheme, Override } from '../types';
import {
  setupToken,
  setupTokens,
  flattenProperties,
  deepExtend,
} from './utils';
import { WebDesignToken } from '../tokens/types/designToken';
import { createComponentCSS } from './createComponentCSS';
import { isString } from '../../utils';
import { createColorPalette } from './createColorPalette';
import { WebTokens } from '../tokens';
import { createAnimationCSS } from './createAnimationCSS';

/**
 * This will be used like `const myTheme = createTheme({})`
 * `myTheme` can then be passed to a Provider or the generated CSS
 * can be passed to a stylesheet at build-time or run-time.
 * const myTheme = createTheme({})
 * const myOtherTheme = createTheme({}, myTheme);
 */
export function createTheme<TokensType extends WebTokens = WebTokens>(
  theme?: Theme<TokensType> | WebTheme,
  DefaultTheme: DefaultTheme | WebTheme = defaultTheme
): WebTheme {
  // merge theme and DefaultTheme to get a complete theme
  // deepExtend is an internal Style Dictionary method
  // that performs a deep merge on n objects. We could change
  // this to another 3p deep merge solution too.
  const mergedTheme = deepExtend<DefaultTheme>([
    {},
    DefaultTheme,
    {
      ...theme,
      components: {},
    },
  ]);

  const { primaryColor, secondaryColor } = mergedTheme;

  // apply primaryColor and secondaryColor if present
  if (isString(primaryColor)) {
    mergedTheme.tokens.colors.primary = createColorPalette({
      keys: Object.keys(mergedTheme.tokens.colors[primaryColor]),
      value: primaryColor,
    });
  }
  if (isString(secondaryColor)) {
    mergedTheme.tokens.colors.secondary = createColorPalette({
      keys: Object.keys(mergedTheme.tokens.colors[secondaryColor]),
      value: secondaryColor,
    });
  }

  // Setting up the tokens. This is similar to what Style Dictionary
  // does. At the end of this, each token should have:
  // - CSS variable name of itself
  // - its value (reference to another CSS variable or raw value)
  const tokens = setupTokens({
    tokens: mergedTheme.tokens,
    setupToken,
  }) as WebTheme['tokens']; // Setting the type here because setupTokens is recursive

  const { breakpoints, name } = mergedTheme;

  // flattenProperties is another internal Style Dictionary function
  // that creates an array of all tokens.
  let cssText =
    `[data-amplify-theme="${name}"] {\n` +
    flattenProperties(tokens)
      .map((token: WebDesignToken) => `${token.name}: ${token.value};`)
      .join('\n') +
    `\n}\n`;

  if (theme?.components) {
    cssText += createComponentCSS(
      name,
      theme.components,
      tokens,
      mergedTheme.breakpoints
    );
  }

  let overrides: Array<Override> = [];

  if (mergedTheme.animations) {
    cssText += createAnimationCSS({
      animations: mergedTheme.animations,
      tokens,
    });
  }

  /**
   * For each override, we setup the tokens and then generate the CSS.
   * This allows us to have one single CSS string for all possible overrides
   * and avoid re-renders in React, but also support other frameworks as well.
   */
  if (mergedTheme.overrides) {
    overrides = mergedTheme.overrides.map((override) => {
      const overrideTokens = setupTokens({
        tokens: override.tokens,
        setupToken,
      });
      const customProperties = flattenProperties(overrideTokens)
        .map((token: WebDesignToken) => `${token.name}: ${token.value};`)
        .join('\n');
      // Overrides can have a selector, media query, breakpoint, or color mode
      // for creating the selector
      if ('selector' in override) {
        cssText += `\n${override.selector} {\n${customProperties}\n}\n`;
      }
      if ('mediaQuery' in override) {
        cssText += `\n@media (${override.mediaQuery}) {
  [data-amplify-theme="${name}"] {
    ${customProperties}
  }
}\n`;
      }
      if ('breakpoint' in override) {
        const breakpoint = mergedTheme.breakpoints.values[override.breakpoint];
        cssText += `\n@media (min-width: ${breakpoint}px) {
  [data-amplify-theme="${name}"] {
    ${customProperties}
  }
}\n`;
      }
      if ('colorMode' in override) {
        cssText += `\n@media (prefers-color-scheme: ${override.colorMode}) {
          [data-amplify-theme="${name}"][data-amplify-color-mode="system"] {
            ${customProperties}
            color-scheme: ${override.colorMode};
          }
        }\n`;
        cssText += `\n[data-amplify-theme="${name}"][data-amplify-color-mode="${override.colorMode}"] {
          ${customProperties}
          color-scheme: ${override.colorMode};
        }\n`;
      }

      return {
        ...override,
        tokens: overrideTokens,
      };
    });
  }

  return {
    tokens,
    breakpoints,
    name,
    cssText,
    containerProps: ({ colorMode } = {}) => {
      return {
        'data-amplify-theme': name,
        'data-amplify-color-mode': colorMode,
      };
    },
    // keep overrides separate from base theme
    // this allows web platforms to use plain CSS scoped to a
    // selector and only override the CSS vars needed. This
    // means we could generate CSS at build-time in a postcss
    // plugin, or do it at runtime and inject the CSS into a
    // style tag.
    // This also allows RN to dynamically switch themes in a
    // provider.
    overrides,
  };
}
