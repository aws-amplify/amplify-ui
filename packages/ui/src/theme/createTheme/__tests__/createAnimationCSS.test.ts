import { createAnimationCSS } from '../createAnimationCSS';
import { createTheme } from '../createTheme';
const { tokens } = createTheme();
describe('@aws-amplify/ui', () => {
  describe('createAnimationCSS', () => {
    it('should work with a theme as a function or an object', () => {
      const css = createAnimationCSS({
        animations: {
          dot: {
            '0%': {
              opacity: 1,
            },
            '100%': {
              opacity: 0,
            },
          },
          example: {
            from: (tokens) => {
              return {
                color: tokens.colors.primary[20],
              };
            },
            to: (tokens) => {
              return {
                color: tokens.colors.primary[40],
              };
            },
          },
          invalid: {
            // @ts-expect-error
            foo: {
              opacity: 1,
            },
            bar: {
              opacity: 0,
            },
          },
        },
        tokens,
      });
      expect(css).toMatchSnapshot();
    });
  });
});
