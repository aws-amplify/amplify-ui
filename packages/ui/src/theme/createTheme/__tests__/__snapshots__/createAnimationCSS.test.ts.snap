// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@aws-amplify/ui createAnimationCSS should work with a theme as a function or an object 1`] = `
"
  @keyframes dot {
    0% {
opacity:1; 
    }
    100% {
opacity:0; 
    }
  }
  @keyframes example {
    from {
color:var(--amplify-colors-primary-20); 
    }
    to {
color:var(--amplify-colors-primary-40); 
    }
  }
  @keyframes invalid {
    foo {
opacity:1; 
    }
    bar {
opacity:0; 
    }
  }"
`;
